generators/java/constants.cjs

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2013-2021 the original author or authors from the JHipster project.
+ *
+ * This file is part of the JHipster project, see https://www.jhipster.tech/
+ * for more information.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const { MAVEN: BUILD_TOOL_MAVEN, MAVEN_DESCRIPTION } = require('../maven/constants.cjs');
+const { GRADLE: BUILD_TOOL_GRADLE, GRADLE_DESCRIPTION } = require('../gradle/constants.cjs');
+
+const JAVA_VERSION = '11';
+const JAVA_COMPATIBLE_VERSIONS = ['1.8', '1.9', '10', '11', '12', '13', '14', '15', '16'];
+const JAVA_APP_VERSION = '0.0.1-SNAPSHOT';
+const JHIPSTER_BOM_VERSION = '7.1.1-SNAPSHOT';
+
+const SRC_MAIN_JAVA_DIR = 'src/main/java/';
+const SRC_MAIN_RESOURCES_DIR = 'src/main/resources/';
+const SRC_TEST_JAVA_DIR = 'src/test/java/';
+
+const BUILD_TOOL = 'buildTool';
```

Comment:
Hmmmm, shouldn't it be in a 'buildTool' constant.js file?

I would have see that kind of file tree structure
```
generator
|_buildTool
   |-constants.js
   |-index.js (helper, common to all vendors)
   |-prompts.js (common to all vendors)
   |_maven
      |-index.js
      |-files.js
      |-prompts.js (vendor specific prompt)
      |_templates
   |_gradle
```
I know it's a bit impacting, but I think it would bring some repeatable guidelines and ease of readability

Reply:
A build-tool generator would be too small and will become too bloated.
And build-tool should be named java-build-tool instead of just build-tool, webpack/angular-cli are js build-tool.
I decided to keep in java generator for this reason.

The tree structure you proposed would require generators with namespace 'jhipster:buildTool:java'.
It's possible but it's not supported by default at yeoman platform.

Reply:
My concern was to avoid having buildTool variable in java generator: from what I understood, the chain is java gen then maven|gradle gen: it would have been more readable to have java constants only on the java generator and buildtool constant in its dedicated generator.
Namespace looks to make sense, even useful in some cases wdyt? example: `search:couchbase` vs `database:couchbase`.
I know that buildtool generator would be thin, but for constants I've no other silver bullet...

Reply:
I am failing to see an implementation that won't require java/build-tool cross-dependency.
generator-java needs to load BUILD_DESTINATION (target for maven, build for gradle).
generator-maven/gradle needs to load JAVA_VERSION, supported java versions, folders from generator-java.

We won't manage to implement only parent->child relationships between generators.
We will need to implement siblings generators java<->maven/gradle, in this case, java is an integration generator, while maven/gradle are specialized generators.

> Namespace looks to make sense, even useful in some cases wdyt? example: `search:couchbase` vs `database:couchbase`.

Can be useful, but in this case it should be `jhipster:spring-boot:search:couchbase` and `jhipster:spring-boot:database:couchbase`.
namespaces can become complicated.

Reply:
Makes perfect sense, thank you for that crystal clear explanation.
Shouldn't it make sense to explicitly separate sibling generator and chained generator, thus splitting the `mixBlueprintGenerator` method in two different `childBlueprintGenerator` and `siblingBlueprintGenerator`.
- a child blueprint can get the properties coming from one of its parent
- a sibling blueprint cannot, in this case a common constant.js and prompt.js file could be specified and common to siblings

Does it make sense?

Reply:
> Shouldn't it make sense to explicitly separate sibling generator and chained generator, thus splitting the `mixBlueprintGenerator` method in two different `childBlueprintGenerator` and `siblingBlueprintGenerator`.

I don't see any reason for that.

> a child blueprint can get the properties coming from one of its parent, a sibling blueprint cannot,

maven generator depends on generator-project-name too, so it depends on the entire generator-java chain.

> in this case a common constant.js and prompt.js file could be specified and common to siblings

- generator-java is already a common generator for gradle and maven.
- maven and gradle don't contain config.js neither options.js.

For the current java/maven/gradle relationship, IMO it's ok as it is now.

We may need to adjust the implementation for more complex generators like spring-boot-(sql, cache, cassandra, mongodb, couchbase, neo4j), spring-boot-web, spring-boot-web-(session, jwt, oauth2).